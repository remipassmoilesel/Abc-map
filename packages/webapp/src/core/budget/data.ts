/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { I18nText, Language } from '@abc-map/shared';
import { BudgetItem } from './BudgetItem';

export const Expenses: BudgetItem[] = [
  {
    id: '2020-1',
    description: hostingLabel(2020),
    value: -633.6,
  },
  {
    id: '2021-1',
    description: hostingLabel(2021),
    value: -633.6,
  },
  {
    id: '2021-2',
    description: [
      { language: Language.French, text: 'Prestation UX 2021' },
      { language: Language.English, text: 'UX service 2021' },
    ],
    value: -95.05,
  },
  {
    id: '2022-1',
    description: hostingLabel(2022),
    value: -633.6,
  },
  {
    id: '2023-1',
    description: hostingLabel(2023),
    value: -633.6,
  },
];

export const Income: BudgetItem[] = [
  {
    id: '2020-1',
    description: donationsLabel(2021),
    value: 85.59,
  },
  {
    id: '2021-1',
    description: donationsLabel(2021),
    value: 88.94,
  },
  {
    id: '2022-1',
    description: donationsLabel(2022),
    value: 9.16,
  },
  {
    id: '2023-1',
    description: donationsLabel(2023),
    value: 9.36,
  },
];

function donationsLabel(year: number): I18nText[] {
  return [
    { language: Language.French, text: 'Dons ' + year },
    { language: Language.English, text: 'Donations ' + year },
  ];
}

function hostingLabel(year: number): I18nText[] {
  return [
    { language: Language.French, text: 'Hébergement ' + year },
    { language: Language.English, text: 'Hosting ' + year },
  ];
}
