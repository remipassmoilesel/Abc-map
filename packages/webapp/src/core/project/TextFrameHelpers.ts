/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { AbcFile, AbcLayout, AbcSharedView, AbcTextFrame, Logger, TextFrameChild } from '@abc-map/shared';
import { nanoid } from 'nanoid';
import cloneDeep from 'lodash/cloneDeep';

const logger = Logger.get('TextFrameHelpers');

// FIXME: we should cache images per source links
export class TextFrameHelpers {
  public static async extractImagesFromLayouts(input: AbcLayout[]): Promise<[AbcFile<Blob>[], AbcLayout[]]> {
    let images: AbcFile<Blob>[] = [];
    const output: AbcLayout[] = [];
    for (const layout of input) {
      const [imgs, frames] = await this.extractImagesFromFrames(layout.textFrames);
      images = images.concat(imgs);
      output.push({ ...layout, textFrames: frames });
    }

    return [images, output];
  }

  public static async extractImagesFromSharedViews(input: AbcSharedView[]): Promise<[AbcFile<Blob>[], AbcSharedView[]]> {
    let images: AbcFile<Blob>[] = [];
    const output: AbcSharedView[] = [];
    for (const view of input) {
      const [imgs, frames] = await this.extractImagesFromFrames(view.textFrames);
      images = images.concat(imgs);
      output.push({ ...view, textFrames: frames });
    }

    return [images, output];
  }

  private static async extractImagesFromFrames(input: AbcTextFrame[]): Promise<[AbcFile<Blob>[], AbcTextFrame[]]> {
    let images: AbcFile<Blob>[] = [];
    const output: AbcTextFrame[] = [];

    for (const frame of input) {
      const [imgs, content] = await this.extractImagesFromElements(frame.content);
      images = images.concat(imgs);
      output.push({ ...frame, content });
    }

    return [images, output];
  }

  private static async extractImagesFromElements(input: TextFrameChild[]): Promise<[AbcFile<Blob>[], TextFrameChild[]]> {
    let images: AbcFile<Blob>[] = [];
    const output: TextFrameChild[] = [];

    for (const element of input) {
      // Element is an image, we extract blob from URL and replace url by a local path
      if (element && 'type' in element && element.type === 'image') {
        const path = `/images/image-${nanoid()}`;
        const file = await fetch(element.url)
          .then((response) => response.blob())
          .then((blob) => ({ path, content: blob }));

        images.push(file);
        output.push({ ...element, url: path });
      }
      // Element is not an image but have children, we inspect it
      else if ('children' in element) {
        const [imgs, children] = await this.extractImagesFromElements(element.children);
        images = images.concat(imgs);
        output.push({ ...element, children } as TextFrameChild);
      }
      // Element is leaf
      else {
        output.push(element);
      }
    }

    return [images, output];
  }

  public static loadImagesOfLayouts(layouts: AbcLayout[], files: AbcFile<Blob>[]): AbcLayout[] {
    return layouts.map((layout) => ({ ...layout, textFrames: this.loadImagesOfFrames(layout.textFrames, files) }));
  }

  public static loadImagesOfSharedViews(views: AbcSharedView[], files: AbcFile<Blob>[]): AbcSharedView[] {
    return views.map((view) => ({ ...view, textFrames: this.loadImagesOfFrames(view.textFrames, files) }));
  }

  private static loadImagesOfFrames(input: AbcTextFrame[], files: AbcFile<Blob>[]): AbcTextFrame[] {
    return input.map((frame) => ({ ...frame, content: this.loadImagesOfElements(frame.content, files) }));
  }

  private static loadImagesOfElements(input: TextFrameChild[], files: AbcFile<Blob>[]): TextFrameChild[] {
    const output: TextFrameChild[] = [];

    for (const element of input) {
      // Element is an image, we create blob url from file
      // If file does not exist we use original path to create a broken image element
      if (element && 'type' in element && element.type === 'image') {
        const file = files.find((f) => f.path === element.url);
        if (!file) {
          logger.error(`Image with path ${element.url} not found, cannot load it`);
        }

        const path = file ? URL.createObjectURL(file.content) : element.url;
        output.push({ ...element, url: path });
      }
      // Element is not an image but have children, we inspect it
      else if ('children' in element) {
        output.push({ ...element, children: this.loadImagesOfElements(element.children, files) } as TextFrameChild);
      }
      // Element is leaf
      else {
        output.push(element);
      }
    }

    return output;
  }

  public static clone(frame: AbcTextFrame): AbcTextFrame {
    return {
      ...cloneDeep(frame),
      id: nanoid(),
    };
  }
}
