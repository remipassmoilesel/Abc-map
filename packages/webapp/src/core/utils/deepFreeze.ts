/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

export declare type Freezable = { [k: string]: any };

export function deepFreeze<T extends Freezable>(obj: T): T {
  Object.keys(obj).forEach((prop) => {
    if (typeof obj[prop] === 'object' && !Object.isFrozen(obj[prop])) deepFreeze(obj[prop]);
  });
  return Object.freeze(obj);
}
