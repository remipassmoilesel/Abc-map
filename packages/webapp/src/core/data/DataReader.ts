/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { DateTime } from 'luxon';
import { getServices, Services } from '../Services';
import { prefixedTranslation } from '../../i18n/i18n';
import { GpxReader } from './readers/GpxReader';
import { KmlReader } from './readers/KmlReader';
import { ShapefileReader } from './readers/ShapefileReader';
import { GeoJsonReader } from './readers/GeoJsonReader';
import { WmsDefinitionReader } from './readers/WmsDefinitionReader';
import { ReaderImplementation as DataReaderImplem } from './readers/ReaderImplementation';
import { AbcArtefact, AbcFile, AbcProjection, Zipper } from '@abc-map/shared';
import { FileFormat, FileFormats } from './FileFormats';
import { ModalStatus } from '../ui/typings';
import { AddLayersChangeset } from '../history/changesets/layers/AddLayersChangeset';
import { HistoryKey } from '../history/HistoryKey';
import { getArea } from 'ol/extent';
import { XyzDefinitionReader } from './readers/XyzDefinitionReader';
import { WmtsDefinitionReader } from './readers/WmtsDefinitionReader';
import { ReadResult, ReadStatus } from './ReadResult';
import { attributionsVariableExpansion } from '../utils/variableExpansion';
import { WktReader } from './readers/WktReader';
import { TopoJsonReader } from './readers/TopoJsonReader';
import { UiConstants } from '../ui/UiConstants';

const t = prefixedTranslation('DataReader:');

export const readersFactory = (services: Services) => [
  new GpxReader(),
  new KmlReader(),
  new ShapefileReader(),
  new GeoJsonReader(),
  new WktReader(),
  new TopoJsonReader(),
  new WmsDefinitionReader(services.geo, services.modals),
  new WmtsDefinitionReader(services.geo, services.modals),
  new XyzDefinitionReader(services.geo, services.modals),
];

export class DataReader {
  public static create(): DataReader {
    return new DataReader(getServices(), readersFactory(getServices()));
  }

  constructor(private services: Services, private readers: DataReaderImplem[]) {}

  public async read(files: AbcFile<Blob>[], targetProjection: AbcProjection): Promise<ReadResult> {
    if (!files.length) {
      throw new Error('No file provided');
    }

    // First we unzip if needed
    const _files: AbcFile<Blob>[] = [];
    for (const f of files) {
      if (FileFormats.fromPath(f.path) === FileFormat.ZIP) {
        const unzipped = await Zipper.forBrowser().unzip(f.content);
        _files.push(...unzipped);
      } else {
        _files.push(f);
      }
    }

    // Then we create layers from data
    const result: ReadResult = { status: ReadStatus.Succeed, layers: [] };
    for (const reader of this.readers) {
      // We read files
      if (await reader.isSupported(_files)) {
        const r = await reader.read(_files, targetProjection);

        // Read succeed, we store layers
        if (r.status === ReadStatus.Succeed && r.layers) {
          result.layers?.push(...r.layers);
        }

        // Operation failed or canceled, we stop reading
        else {
          return { status: r.status };
        }
      }
    }

    if (!result.layers?.length) {
      result.status = ReadStatus.Failed;
    }

    return result;
  }

  public async importFiles(files: AbcFile<Blob>[]): Promise<ReadResult> {
    const { modals, history, geo } = this.services;

    // We check if files are not too big
    const bigFiles = files.filter((f) => f.content.size >= UiConstants.IMPORT_FILE_SIZE_MAX);
    if (bigFiles.length && (await modals.dataSizeWarning()) === ModalStatus.Canceled) {
      return { status: ReadStatus.Canceled, layers: [] };
    }

    // We read files
    const map = geo.getMainMap();
    const readResult = await this.read(files, map.getProjection());
    const layers = readResult.layers;
    if (readResult.status !== ReadStatus.Succeed || !layers?.length) {
      return { status: readResult.status, layers: [] };
    }

    // We check if layers are not too big
    const bigLayers = layers.filter((lay) => lay.isVector() && lay.getSource().getFeatures().length > UiConstants.FEATURE_PER_LAYER_MAX);
    if (!bigFiles.length && bigLayers.length && (await modals.dataSizeWarning()) === ModalStatus.Canceled) {
      return { status: ReadStatus.Canceled, layers: [] };
    }

    // We add layers
    const hour = DateTime.local().toFormat('HH:mm');
    layers.forEach((lay, i) => lay.setName(t('Import_of', { hour, n: i + 1 })));

    const cs = AddLayersChangeset.create(layers);
    await cs.execute();
    history.register(HistoryKey.Map, cs);

    // We fit view on last one
    const last = layers[layers.length - 1];
    const extent = last.unwrap().getExtent();
    if (extent && getArea(extent)) {
      map.unwrap().getView().fit(extent);
    }

    return { status: ReadStatus.Succeed, layers };
  }

  public async importArtefact(artefact: AbcArtefact): Promise<ReadResult> {
    const { dataStore } = this.services;

    // Download then import files
    const files = await dataStore.downloadFilesFrom(artefact);
    const result = await this.importFiles(files);

    // Add attributions
    result.layers?.forEach((lay) => lay.setAttributions(attributionsVariableExpansion(artefact.attributions)));
    return result;
  }
}
