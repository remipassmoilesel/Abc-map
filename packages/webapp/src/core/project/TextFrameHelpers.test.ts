/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { AbcFile, AbcLayout, AbcSharedView } from '@abc-map/shared';
import { TextFrameHelpers } from './TextFrameHelpers';
import { SinonStub } from 'sinon';
import { mockNanoid, restoreNanoid } from '../utils/test/mock-nanoid';
import { sampleFramesWithBlobUrls, sampleFramesWithPaths } from './TextFrameHelpers.test.data';
import { deepFreeze } from '../utils/deepFreeze';
import { TestHelper } from '../utils/test/TestHelper';

describe('TextFrameHelpers', () => {
  describe('Extract images', () => {
    const fetchStub = fetch as SinonStub;
    let layouts: AbcLayout[];
    let sharedViews: AbcSharedView[];

    beforeEach(() => {
      layouts = deepFreeze([
        {
          ...TestHelper.sampleLayout(),
          id: 'test-layout-1',
          textFrames: sampleFramesWithBlobUrls(),
        },
        {
          ...TestHelper.sampleLayout(),
          id: 'test-layout-2',
          textFrames: [sampleFramesWithBlobUrls()[0]],
        },
      ]);

      sharedViews = deepFreeze([
        {
          ...TestHelper.sampleSharedView(),
          id: 'test-shared-view-1',
          textFrames: sampleFramesWithBlobUrls(),
        },
        {
          ...TestHelper.sampleSharedView(),
          id: 'test-shared-view-2',
          textFrames: [sampleFramesWithBlobUrls()[0]],
        },
      ]);

      mockNanoid();

      fetchStub.callsFake((url) => {
        return Promise.resolve({ blob: () => ({ type: 'blob', url }) });
      });
    });

    afterEach(() => {
      restoreNanoid();
    });

    it('extractImagesFromLayouts()', async () => {
      // Act
      const [images, output] = await TextFrameHelpers.extractImagesFromLayouts(layouts);

      // Assert
      expect(images).toEqual([
        { content: { type: 'blob', url: 'http://blob:image-1:17dfc0d8-82af-4602-b84e-9d139c80a406' }, path: '/images/image-###-FAKE-NANOID-1-###' },
        { content: { type: 'blob', url: 'http://blob:image-2:a31099f5-2a60-47cf-a2e0-9dea8e932e93' }, path: '/images/image-###-FAKE-NANOID-2-###' },
        { content: { type: 'blob', url: 'http://blob:image-1:17dfc0d8-82af-4602-b84e-9d139c80a406' }, path: '/images/image-###-FAKE-NANOID-3-###' },
        { content: { type: 'blob', url: 'http://blob:image-2:a31099f5-2a60-47cf-a2e0-9dea8e932e93' }, path: '/images/image-###-FAKE-NANOID-4-###' },
      ]);

      expect(output).toMatchSnapshot();
    });

    it('extractImagesFromSharedViews()', async () => {
      // Act
      const [images, output] = await TextFrameHelpers.extractImagesFromSharedViews(sharedViews);

      // Assert
      expect(images).toEqual([
        { content: { type: 'blob', url: 'http://blob:image-1:17dfc0d8-82af-4602-b84e-9d139c80a406' }, path: '/images/image-###-FAKE-NANOID-1-###' },
        { content: { type: 'blob', url: 'http://blob:image-2:a31099f5-2a60-47cf-a2e0-9dea8e932e93' }, path: '/images/image-###-FAKE-NANOID-2-###' },
        { content: { type: 'blob', url: 'http://blob:image-1:17dfc0d8-82af-4602-b84e-9d139c80a406' }, path: '/images/image-###-FAKE-NANOID-3-###' },
        { content: { type: 'blob', url: 'http://blob:image-2:a31099f5-2a60-47cf-a2e0-9dea8e932e93' }, path: '/images/image-###-FAKE-NANOID-4-###' },
      ]);

      expect(output).toMatchSnapshot();
    });
  });

  describe('Load images', function () {
    const createObjectUrlStub = URL.createObjectURL as SinonStub;
    let layouts: AbcLayout[];
    let sharedViews: AbcSharedView[];
    let files: AbcFile<Blob>[];

    beforeEach(() => {
      layouts = deepFreeze([
        {
          ...TestHelper.sampleLayout(),
          id: 'test-layout-1',
          textFrames: sampleFramesWithPaths(),
        },
        {
          ...TestHelper.sampleLayout(),
          id: 'test-layout-2',
          textFrames: [sampleFramesWithPaths()[0]],
        },
      ]);

      sharedViews = deepFreeze([
        {
          ...TestHelper.sampleSharedView(),
          id: 'test-shared-view-1',
          textFrames: sampleFramesWithPaths(),
        },
        {
          ...TestHelper.sampleSharedView(),
          id: 'test-shared-view-2',
          textFrames: [sampleFramesWithPaths()[0]],
        },
      ]);

      createObjectUrlStub.callsFake((blob) => 'http://blob:' + blob.name);

      files = [
        { path: '/images/image-IMAGE_1', content: { name: 'IMAGE_1' } as unknown as Blob },
        { path: '/images/image-IMAGE_2', content: { name: 'IMAGE_2' } as unknown as Blob },
      ];
    });

    afterEach(() => {
      createObjectUrlStub.reset();
    });

    it('loadImagesOfLayouts()', async () => {
      // Act
      const output = TextFrameHelpers.loadImagesOfLayouts(layouts, files);

      // Assert
      expect(output).toMatchSnapshot();
    });

    it('loadImagesOfSharedViews()', async () => {
      // Act
      const output = TextFrameHelpers.loadImagesOfSharedViews(sharedViews, files);

      // Assert
      expect(output).toMatchSnapshot();
    });
  });
});
