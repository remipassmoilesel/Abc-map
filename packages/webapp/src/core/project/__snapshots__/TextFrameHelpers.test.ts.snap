// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextFrameHelpers Extract images extractImagesFromLayouts() 1`] = `
Array [
  Object {
    "format": Object {
      "height": 297,
      "id": "A4_PORTRAIT",
      "orientation": "portrait",
      "width": 210,
    },
    "id": "test-layout-1",
    "name": "Sample layout",
    "textFrames": Array [
      Object {
        "content": Array [
          Object {
            "text": "Hello",
          },
          Object {
            "children": Array [
              Object {
                "text": "",
              },
            ],
            "size": 1,
            "type": "image",
            "url": "/images/image-###-FAKE-NANOID-1-###",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [
                          Object {
                            "children": Array [
                              Object {
                                "text": "",
                              },
                            ],
                            "size": 2,
                            "type": "image",
                            "url": "/images/image-###-FAKE-NANOID-2-###",
                          },
                        ],
                        "type": "paragraph",
                      },
                    ],
                    "type": "table-cell",
                  },
                ],
                "type": "table-row",
              },
            ],
            "type": "table",
          },
        ],
        "id": "test-frame-1",
        "position": Object {
          "x": 100,
          "y": 200,
        },
        "size": Object {
          "height": 400,
          "width": 300,
        },
        "style": Object {
          "background": "#FFFFFF",
          "withBorders": true,
          "withShadows": true,
        },
      },
      Object {
        "content": Array [
          Object {
            "text": "Hello 2",
          },
          Object {
            "text": "Hello 3",
          },
        ],
        "id": "test-frame-2",
        "position": Object {
          "x": 500,
          "y": 600,
        },
        "size": Object {
          "height": 800,
          "width": 700,
        },
        "style": Object {
          "background": "#FFFFFF",
          "withBorders": true,
          "withShadows": true,
        },
      },
    ],
    "view": Object {
      "center": Array [
        1.5,
        45.4,
      ],
      "projection": Object {
        "name": "EPSG:3857",
      },
      "resolution": 1000,
      "rotation": 0,
    },
  },
  Object {
    "format": Object {
      "height": 297,
      "id": "A4_PORTRAIT",
      "orientation": "portrait",
      "width": 210,
    },
    "id": "test-layout-2",
    "name": "Sample layout",
    "textFrames": Array [
      Object {
        "content": Array [
          Object {
            "text": "Hello",
          },
          Object {
            "children": Array [
              Object {
                "text": "",
              },
            ],
            "size": 1,
            "type": "image",
            "url": "/images/image-###-FAKE-NANOID-3-###",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [
                          Object {
                            "children": Array [
                              Object {
                                "text": "",
                              },
                            ],
                            "size": 2,
                            "type": "image",
                            "url": "/images/image-###-FAKE-NANOID-4-###",
                          },
                        ],
                        "type": "paragraph",
                      },
                    ],
                    "type": "table-cell",
                  },
                ],
                "type": "table-row",
              },
            ],
            "type": "table",
          },
        ],
        "id": "test-frame-1",
        "position": Object {
          "x": 100,
          "y": 200,
        },
        "size": Object {
          "height": 400,
          "width": 300,
        },
        "style": Object {
          "background": "#FFFFFF",
          "withBorders": true,
          "withShadows": true,
        },
      },
    ],
    "view": Object {
      "center": Array [
        1.5,
        45.4,
      ],
      "projection": Object {
        "name": "EPSG:3857",
      },
      "resolution": 1000,
      "rotation": 0,
    },
  },
]
`;

exports[`TextFrameHelpers Extract images extractImagesFromSharedViews() 1`] = `
Array [
  Object {
    "id": "test-shared-view-1",
    "layers": Array [],
    "scale": Object {
      "x": 30,
      "y": 100,
    },
    "textFrames": Array [
      Object {
        "content": Array [
          Object {
            "text": "Hello",
          },
          Object {
            "children": Array [
              Object {
                "text": "",
              },
            ],
            "size": 1,
            "type": "image",
            "url": "/images/image-###-FAKE-NANOID-1-###",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [
                          Object {
                            "children": Array [
                              Object {
                                "text": "",
                              },
                            ],
                            "size": 2,
                            "type": "image",
                            "url": "/images/image-###-FAKE-NANOID-2-###",
                          },
                        ],
                        "type": "paragraph",
                      },
                    ],
                    "type": "table-cell",
                  },
                ],
                "type": "table-row",
              },
            ],
            "type": "table",
          },
        ],
        "id": "test-frame-1",
        "position": Object {
          "x": 100,
          "y": 200,
        },
        "size": Object {
          "height": 400,
          "width": 300,
        },
        "style": Object {
          "background": "#FFFFFF",
          "withBorders": true,
          "withShadows": true,
        },
      },
      Object {
        "content": Array [
          Object {
            "text": "Hello 2",
          },
          Object {
            "text": "Hello 3",
          },
        ],
        "id": "test-frame-2",
        "position": Object {
          "x": 500,
          "y": 600,
        },
        "size": Object {
          "height": 800,
          "width": 700,
        },
        "style": Object {
          "background": "#FFFFFF",
          "withBorders": true,
          "withShadows": true,
        },
      },
    ],
    "title": "Sample layout",
    "view": Object {
      "center": Array [
        1.5,
        45.4,
      ],
      "projection": Object {
        "name": "EPSG:3857",
      },
      "resolution": 1000,
      "rotation": 0,
    },
  },
  Object {
    "id": "test-shared-view-2",
    "layers": Array [],
    "scale": Object {
      "x": 30,
      "y": 100,
    },
    "textFrames": Array [
      Object {
        "content": Array [
          Object {
            "text": "Hello",
          },
          Object {
            "children": Array [
              Object {
                "text": "",
              },
            ],
            "size": 1,
            "type": "image",
            "url": "/images/image-###-FAKE-NANOID-3-###",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [
                          Object {
                            "children": Array [
                              Object {
                                "text": "",
                              },
                            ],
                            "size": 2,
                            "type": "image",
                            "url": "/images/image-###-FAKE-NANOID-4-###",
                          },
                        ],
                        "type": "paragraph",
                      },
                    ],
                    "type": "table-cell",
                  },
                ],
                "type": "table-row",
              },
            ],
            "type": "table",
          },
        ],
        "id": "test-frame-1",
        "position": Object {
          "x": 100,
          "y": 200,
        },
        "size": Object {
          "height": 400,
          "width": 300,
        },
        "style": Object {
          "background": "#FFFFFF",
          "withBorders": true,
          "withShadows": true,
        },
      },
    ],
    "title": "Sample layout",
    "view": Object {
      "center": Array [
        1.5,
        45.4,
      ],
      "projection": Object {
        "name": "EPSG:3857",
      },
      "resolution": 1000,
      "rotation": 0,
    },
  },
]
`;

exports[`TextFrameHelpers Load images loadImagesOfLayouts() 1`] = `
Array [
  Object {
    "format": Object {
      "height": 297,
      "id": "A4_PORTRAIT",
      "orientation": "portrait",
      "width": 210,
    },
    "id": "test-layout-1",
    "name": "Sample layout",
    "textFrames": Array [
      Object {
        "content": Array [
          Object {
            "text": "Hello",
          },
          Object {
            "children": Array [
              Object {
                "text": "",
              },
            ],
            "size": 1,
            "type": "image",
            "url": "http://blob:IMAGE_1",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [
                          Object {
                            "children": Array [
                              Object {
                                "text": "",
                              },
                            ],
                            "size": 2,
                            "type": "image",
                            "url": "http://blob:IMAGE_2",
                          },
                        ],
                        "type": "paragraph",
                      },
                    ],
                    "type": "table-cell",
                  },
                ],
                "type": "table-row",
              },
            ],
            "type": "table",
          },
        ],
        "id": "test-frame-1",
        "position": Object {
          "x": 100,
          "y": 200,
        },
        "size": Object {
          "height": 400,
          "width": 300,
        },
        "style": Object {
          "background": "#FFFFFF",
          "withBorders": true,
          "withShadows": true,
        },
      },
      Object {
        "content": Array [
          Object {
            "text": "Hello 2",
          },
          Object {
            "text": "Hello 3",
          },
        ],
        "id": "test-frame-2",
        "position": Object {
          "x": 500,
          "y": 600,
        },
        "size": Object {
          "height": 800,
          "width": 700,
        },
        "style": Object {
          "background": "#FFFFFF",
          "withBorders": true,
          "withShadows": true,
        },
      },
    ],
    "view": Object {
      "center": Array [
        1.5,
        45.4,
      ],
      "projection": Object {
        "name": "EPSG:3857",
      },
      "resolution": 1000,
      "rotation": 0,
    },
  },
  Object {
    "format": Object {
      "height": 297,
      "id": "A4_PORTRAIT",
      "orientation": "portrait",
      "width": 210,
    },
    "id": "test-layout-2",
    "name": "Sample layout",
    "textFrames": Array [
      Object {
        "content": Array [
          Object {
            "text": "Hello",
          },
          Object {
            "children": Array [
              Object {
                "text": "",
              },
            ],
            "size": 1,
            "type": "image",
            "url": "http://blob:IMAGE_1",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [
                          Object {
                            "children": Array [
                              Object {
                                "text": "",
                              },
                            ],
                            "size": 2,
                            "type": "image",
                            "url": "http://blob:IMAGE_2",
                          },
                        ],
                        "type": "paragraph",
                      },
                    ],
                    "type": "table-cell",
                  },
                ],
                "type": "table-row",
              },
            ],
            "type": "table",
          },
        ],
        "id": "test-frame-1",
        "position": Object {
          "x": 100,
          "y": 200,
        },
        "size": Object {
          "height": 400,
          "width": 300,
        },
        "style": Object {
          "background": "#FFFFFF",
          "withBorders": true,
          "withShadows": true,
        },
      },
    ],
    "view": Object {
      "center": Array [
        1.5,
        45.4,
      ],
      "projection": Object {
        "name": "EPSG:3857",
      },
      "resolution": 1000,
      "rotation": 0,
    },
  },
]
`;

exports[`TextFrameHelpers Load images loadImagesOfSharedViews() 1`] = `
Array [
  Object {
    "id": "test-shared-view-1",
    "layers": Array [],
    "scale": Object {
      "x": 30,
      "y": 100,
    },
    "textFrames": Array [
      Object {
        "content": Array [
          Object {
            "text": "Hello",
          },
          Object {
            "children": Array [
              Object {
                "text": "",
              },
            ],
            "size": 1,
            "type": "image",
            "url": "http://blob:IMAGE_1",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [
                          Object {
                            "children": Array [
                              Object {
                                "text": "",
                              },
                            ],
                            "size": 2,
                            "type": "image",
                            "url": "http://blob:IMAGE_2",
                          },
                        ],
                        "type": "paragraph",
                      },
                    ],
                    "type": "table-cell",
                  },
                ],
                "type": "table-row",
              },
            ],
            "type": "table",
          },
        ],
        "id": "test-frame-1",
        "position": Object {
          "x": 100,
          "y": 200,
        },
        "size": Object {
          "height": 400,
          "width": 300,
        },
        "style": Object {
          "background": "#FFFFFF",
          "withBorders": true,
          "withShadows": true,
        },
      },
      Object {
        "content": Array [
          Object {
            "text": "Hello 2",
          },
          Object {
            "text": "Hello 3",
          },
        ],
        "id": "test-frame-2",
        "position": Object {
          "x": 500,
          "y": 600,
        },
        "size": Object {
          "height": 800,
          "width": 700,
        },
        "style": Object {
          "background": "#FFFFFF",
          "withBorders": true,
          "withShadows": true,
        },
      },
    ],
    "title": "Sample layout",
    "view": Object {
      "center": Array [
        1.5,
        45.4,
      ],
      "projection": Object {
        "name": "EPSG:3857",
      },
      "resolution": 1000,
      "rotation": 0,
    },
  },
  Object {
    "id": "test-shared-view-2",
    "layers": Array [],
    "scale": Object {
      "x": 30,
      "y": 100,
    },
    "textFrames": Array [
      Object {
        "content": Array [
          Object {
            "text": "Hello",
          },
          Object {
            "children": Array [
              Object {
                "text": "",
              },
            ],
            "size": 1,
            "type": "image",
            "url": "http://blob:IMAGE_1",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [
                          Object {
                            "children": Array [
                              Object {
                                "text": "",
                              },
                            ],
                            "size": 2,
                            "type": "image",
                            "url": "http://blob:IMAGE_2",
                          },
                        ],
                        "type": "paragraph",
                      },
                    ],
                    "type": "table-cell",
                  },
                ],
                "type": "table-row",
              },
            ],
            "type": "table",
          },
        ],
        "id": "test-frame-1",
        "position": Object {
          "x": 100,
          "y": 200,
        },
        "size": Object {
          "height": 400,
          "width": 300,
        },
        "style": Object {
          "background": "#FFFFFF",
          "withBorders": true,
          "withShadows": true,
        },
      },
    ],
    "title": "Sample layout",
    "view": Object {
      "center": Array [
        1.5,
        45.4,
      ],
      "projection": Object {
        "name": "EPSG:3857",
      },
      "resolution": 1000,
      "rotation": 0,
    },
  },
]
`;
